### PR TITLE
Fix excessive memory use (#405)

### LINKED ISSUE #405: ADRIAmod uses excessive amounts of memory during runs
ADRIAmod uses up to 99% memory when doing larger numbers of runs.

The method used to run scenarios is to create an anonymous function (i.e., a closure) to which we pass in each row in the scenario spec. `pmap` is used regardless of whether parallelism is active or not (where a single worker is used for single core use cases).

A named tuple of matrices used as intermediate caches throughout a simulation is used to reduce memory allocation. The size of these matrices are dependent on the number of species/locations represented and can be quite large.

Additionally, it seems that `pmap` will naively serialize data to send to each worker for every job. Using `CachingPool()` avoids this (as indicated in: https://github.com/JuliaLang/julia/pull/33892) and is set to be the default behavior (in Julia v1.10 I believe). This means currently these large cache matrices get reserialized for every scenario run. The Julia's garbage collector doesn't seem to clean these up very often, so they hang around, which can then lead to out-of-memory issues.

I have successfully run 8192 scenarios with just `CachingPool()` active (comparably high memory use, but Julia does not crash), and am trying creating the cache data structures within each scenario run to avoid their serialization.

The reserialization might explain why @Rosejoycrocker sees Julia crash even for single core runs (as `pmap()` is used regardless of whether runs are done with single or multiple cores).

A smarter strategy may be needed, where the regular `map()` is used for single core runs. The advantage is that in such cases the cache is created once and reused for every run, reducing memory use in single-core contexts.

### PR DESCRIPTION
**Pull request by ConnectedSystems**
Closes #405 

Includes a number of changes to avoid Julia kernel crashing when running a high number of scenarios.

- Only uses `pmap` in workloads that require multiprocessing, otherwises uses `map`
- Does not create a reusable cache structure when using multiple cores (but does for single core workloads)
- Explicit use of `CachingPool()` to avoid re-serialization of data structures to workers (default behavior in upcoming Julia v1.10 release)
- Forces garbage collector (gc) to run if 95% of available memory is in use. Forcing the gc to run too often can increase runtimes and lead to other issues so I've opted to only run this when absolutely needed

In addition to the above, a key change resolving crashes in multi-processing contexts was (circa L 271 in `scenarios.jl`):

```julia
# dhw_scen::Matrix{Float64} = @view domain.dhw_scens[:, :, dhw_idx]

# Remove the use of array `view` and copy the data for the given DHW scenario instead:
dhw_scen::Matrix{Float64} = copy(domain.dhw_scens[:, :, dhw_idx])
```

I don't have a concrete explanation for this other than I suspect `domain` may be shared across workers and so multiple workers attempting to interact with a "big" object type like `Domain` may have led to issues...

@Rosejoycrocker @Zapiano I'm tagging you both in as reviewers for this, but only need approval from one of you. Just be aware of the changes here please.
 
I have successfully run 8,192 scenarios both using a single core (using the `debug=true` option) and multiple cores (i.e., two).

This is not a concrete fix: you should see large amounts of memory use still but at least Julia itself does not crash and the runs complete.

### CODE REVIEW

**Comment by Rosejoycrocker** [comment]
Hey @ConnectedSystems , should I also try running 8196 scenarios for testing ?

**Comment by ConnectedSystems** [comment]
It would be good if you did, yes.

**Comment by Rosejoycrocker** [comment]
I'm still getting crashes for 8192 runs. I updated ADRIA and my sandbox environment- is there anything else I need to do?

**Comment by ConnectedSystems** [comment]
Is Julia crashing or is it the run that crashes?

**Comment by Rosejoycrocker** [comment]
Julia crashes and I get the exception access violation error

**Comment by ConnectedSystems** [comment]
Could you post the error message in full here please?

**Comment by Rosejoycrocker** [comment]
Sorry it's huge but here it is:
```
Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x7ff9459453a5 -- getindex at .\essentials.jl:14 [inlined]
macro expansion at .\multidimensional.jl:889 [inlined]
macro expansion at .\cartesian.jl:64 [inlined]
_unsafe_getindex! at .\multidimensional.jl:884 [inlined]
_unsafe_getindex at .\multidimensional.jl:875
in expression starting at C:\Users\rcrocker\Documents\GitHub\ADRIA.jl\sandbox\mcda_alg_sensitivity.jl:49
getindex at .\essentials.jl:14 [inlined]
macro expansion at .\multidimensional.jl:889 [inlined]
macro expansion at .\cartesian.jl:64 [inlined]
_unsafe_getindex! at .\multidimensional.jl:884 [inlined]
_unsafe_getindex at .\multidimensional.jl:875
_getindex at .\multidimensional.jl:861 [inlined]
getindex at .\abstractarray.jl:1294 [inlined]
run_model at C:\Users\rcrocker\Documents\GitHub\ADRIA.jl\src\scenario.jl:514
run_scenario at C:\Users\rcrocker\Documents\GitHub\ADRIA.jl\src\scenario.jl:185
unknown function (ip: 00000279d14ae182)
#99 at C:\Users\rcrocker\Documents\GitHub\ADRIA.jl\src\scenario.jl:137
#56 at C:\Users\rcrocker\.julia\packages\ProgressMeter\sN2xr\src\ProgressMeter.jl:1016
iterate at .\generator.jl:47 [inlined]
collect_to! at .\array.jl:840
collect_to_with_first! at .\array.jl:818 [inlined]
collect at .\array.jl:792
unknown function (ip: 00000279d1479796)
map at .\abstractarray.jl:3289
unknown function (ip: 00000279d147647a)
jl_apply at C:/workdir/src\julia.h:1879 [inlined]
do_apply at C:/workdir/src\builtins.c:730
macro expansion at C:\Users\rcrocker\.julia\packages\ProgressMeter\sN2xr\src\ProgressMeter.jl:1015 [inlined]
macro expansion at .\task.jl:476 [inlined]
macro expansion at C:\Users\rcrocker\.julia\packages\ProgressMeter\sN2xr\src\ProgressMeter.jl:1014 [inlined]
macro expansion at .\task.jl:476 [inlined]
#progress_map#53 at C:\Users\rcrocker\.julia\packages\ProgressMeter\sN2xr\src\ProgressMeter.jl:1007
jl_apply at C:/workdir/src\julia.h:1879 [inlined]
do_apply at C:/workdir/src\builtins.c:730
progress_map at C:\Users\rcrocker\.julia\packages\ProgressMeter\sN2xr\src\ProgressMeter.jl:999
#run_scenarios#97 at C:\Users\rcrocker\Documents\GitHub\ADRIA.jl\src\scenario.jl:146
run_scenarios at C:\Users\rcrocker\Documents\GitHub\ADRIA.jl\src\scenario.jl:68 [inlined]
#run_scenarios#96 at C:\Users\rcrocker\Documents\GitHub\ADRIA.jl\src\scenario.jl:66 [inlined]
run_scenarios at C:\Users\rcrocker\Documents\GitHub\ADRIA.jl\src\scenario.jl:65
jl_apply at C:/workdir/src\julia.h:1879 [inlined]
do_call at C:/workdir/src\interpreter.c:126
eval_value at C:/workdir/src\interpreter.c:226
eval_stmt_value at C:/workdir/src\interpreter.c:177 [inlined]
eval_body at C:/workdir/src\interpreter.c:624
jl_interpret_toplevel_thunk at C:/workdir/src\interpreter.c:762
jl_toplevel_eval_flex at C:/workdir/src\toplevel.c:912
jl_toplevel_eval_flex at C:/workdir/src\toplevel.c:856
ijl_toplevel_eval at C:/workdir/src\toplevel.c:921 [inlined]
ijl_toplevel_eval_in at C:/workdir/src\toplevel.c:971
eval at .\boot.jl:370 [inlined]
include_string at .\loading.jl:1903
_include at .\loading.jl:1963
include at .\client.jl:478
unknown function (ip: 00000279f9830406)
jl_apply at C:/workdir/src\julia.h:1879 [inlined]
do_call at C:/workdir/src\interpreter.c:126
eval_value at C:/workdir/src\interpreter.c:226
eval_stmt_value at C:/workdir/src\interpreter.c:177 [inlined]
eval_body at C:/workdir/src\interpreter.c:624
jl_interpret_toplevel_thunk at C:/workdir/src\interpreter.c:762
jl_toplevel_eval_flex at C:/workdir/src\toplevel.c:912
jl_toplevel_eval_flex at C:/workdir/src\toplevel.c:856
eval_body at C:/workdir/src\interpreter.c:568
jl_interpret_toplevel_thunk at C:/workdir/src\interpreter.c:762
jl_toplevel_eval_flex at C:/workdir/src\toplevel.c:912
jl_toplevel_eval_flex at C:/workdir/src\toplevel.c:856
ijl_toplevel_eval at C:/workdir/src\toplevel.c:921 [inlined]
ijl_toplevel_eval_in at C:/workdir/src\toplevel.c:971
eval at .\boot.jl:370 [inlined]
eval_user_input at C:\workdir\usr\share\julia\stdlib\v1.9\REPL\src\REPL.jl:153
repl_backend_loop at C:\workdir\usr\share\julia\stdlib\v1.9\REPL\src\REPL.jl:249
#start_repl_backend#46 at C:\workdir\usr\share\julia\stdlib\v1.9\REPL\src\REPL.jl:234
start_repl_backend at C:\workdir\usr\share\julia\stdlib\v1.9\REPL\src\REPL.jl:231
#run_repl#59 at C:\workdir\usr\share\julia\stdlib\v1.9\REPL\src\REPL.jl:379
run_repl at C:\workdir\usr\share\julia\stdlib\v1.9\REPL\src\REPL.jl:365
jfptr_run_repl_61185.clone_1 at C:\Users\rcrocker\.julia\juliaup\julia-1.9.2+0.x64.w64.mingw32\lib\julia\sys.dll (unknown line)
#1017 at .\client.jl:421
exec_options at .\client.jl:322
_start at .\client.jl:522
jfptr__start_47602.clone_1 at C:\Users\rcrocker\.julia\juliaup\julia-1.9.2+0.x64.w64.mingw32\lib\julia\sys.dll (unknown line)
jl_apply at C:/workdir/src\julia.h:1879 [inlined]
true_main at C:/workdir/src\jlapi.c:573
jl_repl_entrypoint at C:/workdir/src\jlapi.c:717
mainCRTStartup at C:/workdir/cli\loader_exe.c:59
BaseThreadInitThunk at C:\WINDOWS\System32\KERNEL32.DLL (unknown line)
RtlUserThreadStart at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
Allocations: 15658444760 (Pool: 15648644664; Big: 9800096); GC: 5154```

**Comment by Rosejoycrocker** [comment]
I also forgot to say in the meeting- when I tried running on the remote desktop I got a similar exception access error so I  don't think it's my machine

**Comment by ConnectedSystems** [comment]
I'm not seeing this happening, but could there are other issues and I was just lucky in my three runs.

Could you pull and try again?

**Comment by Rosejoycrocker** [comment]
I tried pulling again (it seems like maybe GitKraken was slow to update changes?) but now I'm getting this error when I try to precompile ADRIA:

```
julia> using ADRIA
[ Info: Precompiling ADRIA [7dc409a7-fbe5-4c9d-b3e2-b0c19a6ba600]
ERROR: LoadError: UndefVarError: `wave_scens` not defined
Stacktrace:
  [1] run_model(domain::ADRIA.ADRIADomain{NamedDims.NamedDimsArray{(:Source, :Receiving), Float64, 2, AxisKeys.KeyedArray{Float64, 2, SparseArrays.SparseMatrixCSC{Float64, Int64}, Tuple{Vector{String}, Vector{String}}}}, NamedDims.NamedDimsArray{(:species, :sites), Float32, 2, AxisKeys.KeyedArray{Float32, 2, Matrix{Float32}, Tuple{UnitRange{Int64}, Vector{String}}}}, Vector{Int64}, DataFrames.DataFrame, String, Vector{Float64}, Vector{String}, NamedDims.NamedDimsArray{(:timesteps, :sites, :scenarios), Float32, 3, AxisKeys.KeyedArray{Float32, 3, Array{Float32, 3}, Tuple{UnitRange{Int64}, Vector{String}, UnitRange{Int64}}}}, NamedDims.NamedDimsArray{(:timesteps, :sites, :scenarios), Float32, 3, AxisKeys.KeyedArray{Float32, 3, Array{Float32, 3}, Tuple{UnitRange{Int64}, Vector{String}, UnitRange{Int64}}}}, Matrix{Float64}}, param_set::NamedDims.NamedDimsArray{(:factors,), Any, 1, AxisKeys.KeyedArray{Any, 1, Vector{Any}, Base.RefValue{Vector{String}}}}, corals::DataFrames.DataFrame, cache::NamedTuple{(:sf, :fec_all, :fec_scope, :dhw_step, :cov_tmp, :depth_coeff, :site_area, :TP_data, :waves), Tuple{Matrix{Float64}, Matrix{Float64}, Matrix{Float64}, Vector{Float64}, Matrix{Float64}, Vector{Float64}, Matrix{Float64}, Matrix{Float64}, Array{Float64, 3}}})
    @ ADRIA C:\Users\rcrocker\Documents\GitHub\ADRIA.jl\src\scenario.jl:464 ...
```

**Comment by ConnectedSystems** [comment]
Could you try pulling again please - sorry, it was a typo in the variable name. Issue with doing work while in a meeting.

**Comment by Rosejoycrocker** [comment]
That's all good, I'll pull again :)

**Comment by Rosejoycrocker** [comment]
Thanks Takuya, it's running now. I'll just make sure it doesn't crash then I'll review and merge

**Comment by Rosejoycrocker** [comment]
It's running without crashing and looks good to me.

**Review by Rosejoycrocker — APPROVED** [review]

**Comment by Rosejoycrocker** [comment]
Should I wait for Pedro's review or just merge?

**Comment by Zapiano** [comment]
> Should I wait for Pedro's review or just merge?

@Rosejoycrocker you can go ahead and merge! I am finishing setting up my computer and I don't want to be a blocker.

**Comment by Rosejoycrocker** [comment]
Thanks @Zapiano !

File: src/scenario.jl [hunk 1]
```diff
@@ -89,36 +89,64 @@ function run_scenarios(param_df::DataFrame, domain::Domain, RCP::Vector{String};
...
-        target_rows = findall(scenarios_matrix("RCP") .== parse(Float64, rcp))
-        if show_progress
-            @showprogress run_msg 4 pmap(func, zip(target_rows, eachrow(scenarios_matrix[target_rows, :])))
-        else
-            pmap(func, zip(target_rows, eachrow(scenarios_matrix[target_rows, :])))
+    if parallel
+        for rcp in RCP
+            run_msg = "Running $(nrow(param_df)) scenarios for RCP $rcp"
+
+            # Switch RCPs so correct data is loaded
```
[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/scenario.jl:123]: FYI @DanTanAtAims I think Pedro is correct here, and goes to the issue you found yesterday

[the remark this answers is not in the document]
Comment (DanTanAtAims) [reply src/scenario.jl:123]: Yes, in the draft pull request which I'm still working I've moved the `switch_RCP!` back to where Pedro noted above so as to only load RCPs once, avoiding duplicate loads.

**Pull request merged**